spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -483,16 +483,35 @@ def get_editorstack(self):
             return editor.get_current_editorstack()
         raise RuntimeError('No editorstack found.')
 
-    def handle_save_files(self):
-        """Save the open files."""
-        self.get_editorstack().save()
+    def _handle_get_file_code(self, filename):
+        """
+        Return the bytes that compose the file.
+
+        Bytes are returned instead of str to support non utf-8 files.
+        """
+        editorstack = self.get_editorstack()
+        editorstack.save_all(save_new_files=False)
+        editor = self.get_editor(filename)
+
+        if editor is None:
+            # load from file
+            mode = 'rb'
+            if PY2:
+                mode = 'r'
+            with open(filename, mode) as f:
+                return f.read()
```

Comment:
Instead of reading the file from memory here, we could raise an error and let the kernel load the file (Which it already does if this function raises an error).

Reply:
I don't understand. When can this case happen?

Reply:
Instead of reading the file from memory, the kernel asks the frontend for the file content, and it this doesn't work, loads it from memory. In the case where the file is not open in the frontend, the frontend can either:
 - Raise an error to notify the kernel that the file is not open, so the kernel can load from memory
 - load the file from memory and send this content.
I like the first option better to avoid code duplication and having to send large files over the comm without reason.

Reply:
What if the kernel lives in a remote server and tries to open a file there with the same name and location but that it doesn't correspond to the one the user wants to really run?

Reply:
> By the way, the test failure on Linux and Python 2 look legit (the same test fails every time). So please take a look at that.

I think this is fixed in qtconsole#375

Reply:
Should the file be with the frontend or with the kernel? IN the current master the file lives with the kernel. Should that change?

Reply:
> Should the file be with the frontend or with the kernel?

I think that a basic user would expect the 'F5' shortcut to run the file they are looking at in Spyder regardless of what is on the remote file system. Advanced users would be surprised that it works at all.

Would it simplify anything if the `runfile` could only run file that are open in spyder and if user want to run a file from the file system they could use ipython's %run magic.

We could even change the name of `runfile`

Reply:
> Would it simplify anything if the runfile could only run file that are open in spyder and if user want to run a file from the file system they could use ipython's %run magic.

> We could even change the name of runfile

Or that runfile would open the file in the editor. (As could runcell). This would make running things from history easier. Also it could be a bit surprising if calling runfile gives you:
`File script.py not open in the editor`. It is already a bit surprising that this happens with runcell.

I would say that `runfile` should be used to run things from the frontend side, and `%run` from the kernel side. The error message could say: "the editor can not open `/path/on/kernel/machine.py`" so the problem would be obvious to the advanced user.